torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -441,19 +443,100 @@ void ONNXSetDynamicInputShape(
   }
 }
 
+bool HasSequenceTypeOutput(Node* node) {
+  if (node->kind() == ::c10::onnx::SplitToSequence ||
+      node->kind() == ::c10::onnx::SequenceInsert ||
+      node->kind() == ::c10::onnx::SequenceEmpty ||
+      node->kind() == ::c10::onnx::SequenceErase ||
+      node->kind() == ::c10::onnx::SequenceConstruct)
+    return true;
+  return false;
+}
+
+void ONNXUpdateTypeFromTensor(
+    Value* graph_output,
+    at::Tensor output,
+    bool onnx_shape_inference) {
+  if (onnx_shape_inference) {
+    graph_output->setType(
+        MergeInferredType(TensorType::create(output), graph_output->type()));
+  } else {
+    graph_output->inferTypeFrom(output);
+  }
+}
+
 void ONNXAssignOutputShape(
     std::shared_ptr<Graph>& graph,
     at::ArrayRef<at::Tensor> outputs,
+    const python::IODescriptor& desc,
     bool onnx_shape_inference) {
-  TORCH_INTERNAL_ASSERT(graph->outputs().size() == outputs.size());
-  for (size_t i = 0; i < outputs.size(); ++i) {
-    if (onnx_shape_inference) {
-      graph->outputs()[i]->setType(MergeInferredType(
-          TensorType::create(outputs[i]), graph->outputs()[i]->type()));
-    } else {
-      graph->outputs()[i]->inferTypeFrom(outputs[i]);
+  size_t outputs_index = 0;
+
+  PyObject* py_obj = unflatten(outputs, desc);
+
+  TORCH_INTERNAL_ASSERT(PyTuple_Check(py_obj));
+
+  for (size_t i = 0; i < PyTuple_GET_SIZE(py_obj); ++i) {
+    PyObject* elem = PyTuple_GET_ITEM(py_obj, i);
+
+    if (PyList_Check(elem)) {
+      size_t list_len = PyList_GET_SIZE(elem);
+      TORCH_INTERNAL_ASSERT(HasSequenceTypeOutput(graph->outputs()[i]->node()))
+      if (list_len > 0) {
+        auto& var =
+            reinterpret_cast<THPVariable*>(PyList_GET_ITEM(elem, 0))->cdata;
+        for (size_t j = 1; j < list_len; ++j) {
+          PyObject* list_elem = PyList_GET_ITEM(elem, j);
+          TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+          auto& new_var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+          TORCH_CHECK(
+              var.scalar_type() == new_var.scalar_type(),
+              "Unsupported sequence type in model outputs. ONNX supports sequences of elements of the same data type.");
+        }
+        graph->outputs()[i]->setType(ListType::create(TensorType::create(var)));
```

Comment:
i'm still unsure here. Can we change this PR to only update scalarType for existing graph output?

TL;DR

Current code will be overwriting graph output type/shape with the one inferred from the first tensor in the list. I think this may not be accurate. 

The most accurate solution would be to go over all the `list_elem`, find the common shape (same logic as in ONNX).

e.g.
`a:[3, 4], b:[4, 4] => c:[None, 4]`
`a:[3, 4, 5], b:[4, 4] => c:None`

And at last merge with existing shape in graph (this is also tricky).

On the other hand i think it's ok not to block this PR with the issue above since several shape inference infras are not ready and are being improved at the moment. We will need to revisit this as we enable more and more sequence use cases.